shelley/chain-and-ledger/executable-spec/src/Updates.hs

```diff
@@ -158,10 +163,34 @@ updatePPup
   -> PPUpdate hashAlgo dsignAlgo
 updatePPup (PPUpdate pup0') (PPUpdate pup1') = PPUpdate (pup1' ∪ pup0')
 
-newAVs :: Applications -> Map.Map Slot Applications -> Applications
-newAVs avs favs = if not $ Map.null favs
-  then let maxSlot = maximum $ Map.keys favs in favs Map.! maxSlot  -- value exists because maxSlot is in keys
-  else avs
+-- | This is just an example and not neccessarily how we will actually validate names
+apNameValid :: ApName -> Bool
+apNameValid (ApName an) = all isAscii cs && length cs <= 12
+  where cs = unpack an
+
+type Favs = Map.Map Slot Applications
+
+maxVer :: ApName -> Applications -> Favs -> (ApVer, Metadata)
+maxVer an avs favs =
+  maximum $ vs an avs : fmap (vs an) (Map.elems favs)
+    where
+      vs n (Applications as) =
+        Map.foldr max (ApVer 0, Metadata) (Map.filterWithKey (\k _ -> k == n) as)
+
+svCanFollow :: Applications -> Favs -> (ApName, (ApVer, Metadata)) -> Bool
+svCanFollow avs favs (an, (ApVer v, _)) = v == 1 + m
+  where (ApVer m) = fst $ maxVer an avs favs
+
+allNames :: Applications -> Favs -> Set ApName
+allNames (Applications avs) favs =
```

Comment:
Relation to the rescue...

`Prelude.foldr
	    (\(Applications fav) acc -> acc ∪ dom fav) (dom avs) (range favs)`